### ISSUE TITLE
symbolic links and docker 

### ISSUE THREAD

**Issue by pavlis** [mspass-team/mspass#335]
The same tests with pwmig that led to issue #332 showed error logs from that test with this content:
```
[LogData({'job_id': 0, 'p_id': 2041, 'algorithm': 'MsPASSError', 'message': 'fread_data_from_file:  Open failed on file simdata/simdata_22', 'badness': ErrorSeverity.Invalid})]
```

I could not understand why open had failed on that file when interacting the parent file system with the unix shell.   However, when I went to the terminal in the jupyter notebook window the problem became obvious.  In the test setup I was using I had created a symbolic link and the directory "simdata" was actually a symbolic link to another location in the file system.   A common trick, BUT because I was running docker telling it to mount the file system as /home in the "." directory where simdata was referenced that data is not visible from the container.   

This is likely a well known issue with docker containers BUT it is likely to bite anyone and create mysterious results.   The use of symbolic links is very common in data management to reduce the need to move large quantities of data around and is sometimes essential to stage a large data set with files spread around in multiple file systems.   I think there are three important consequences of this that require followup:

1. We absolutely need to add this warning multiple times in the documentation.   I recommend we do this in at least two ways.  First, we really need to start a "best practices" page for the user manual anyway.   This is an obvious topic for a section of such a page.  Second, a warning about this issue should be added to the user manual and wiki sections on the use of docker.  Presume, by the way, singularity would have this same issue.   
2. I think we need to add some logic in the `read_data` method of `Database` for file-based reads to test all "dir" entries with the python os.path module function `islink`.   I just tested it with the example above and confirm it does detect the file is a link.  I presume, but have not yet found a good example, that one can then verify if the link resolves somehow.   I can't think of an example where I'd know for  sure that could ever happen in this situation so it may be best to throw an exception when this situation is encountered.  That decision is context dependent. 
3. I think we need to add a test to our [Database.verify](https://www.mspass.org/python_api/mspasspy.db.html#module-mspasspy.db.database) method.  That test would:  (a) subset a wf collection to documents with data defined by files and (b) test all dir entries in documents found for symbolic links.

**Comment by wangyinz** [comment]
Yeah, that's indeed not obvious at all. For `read_data`, I think the fix I just merged addressed the problem already. If the data file is not available, it will now error out instead of silently returning killed objects. This should be a better approach than just verifying the symbolic links. In fact, invalid symbolic links is just one way to mess up the read, and there are a lot of different ways to have files unavailable, so we are really just defining the behavior for unavailable files. This behavior is also consistent with our readers for other formats/sources.

**Comment by pavlis** [comment]
Good to know the new fix will at least prevent such mysterious behavior.

With that suggestion a change to a new function needed for the Database.verify method:   we need a method that will open for read and immediately close all document in a collection defined with files as the storage method.  An option will be needed, come to think of it, for a filtering query to not require testing a full collection.  A database could contain multiple datasets or multiple saves and one might not want or need to test them all.   A process, by the way, that could be very time consuming on a large data set.  Might want to have a parallel option that would run the verify with dask/spark.   

Another thought:   the work might be reduced for some cases with a "distinct" clause in a query for dir/dfile when data are (correctly) assembled in gathers to reduce the number of files.

**Comment by wangyinz** [comment]
Addressed in #766: the Docker and development documentation now explains that symlink targets must also be bind-mounted and that waveform metadata must use the container-visible path. Closing as completed.
